GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h"
+
+@interface GULMulticastAppDelegate () <GULMulticastAppDelegateProtocol> {
+  NSMutableArray<id> *_interceptors;
+  id<GULApplicationDelegate> _defaultAppDelegate;
+}
+@end
+
+@implementation GULMulticastAppDelegate
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _interceptors = [[NSMutableArray alloc] init];
+  }
+  return self;
+}
+
+- (instancetype)initWithAppDelegate:(id<GULApplicationDelegate>)delegate {
+  self = [super init];
+  if (self) {
+    _interceptors = [[NSMutableArray alloc] init];
+    [_interceptors addObject:delegate];
+    _defaultAppDelegate = delegate;
+  }
+  return self;
+}
+
++ (id<GULMulticastAppDelegateProtocol>)multicastDelegate {
+  id<UIApplicationDelegate> appDelegate = [UIApplication sharedApplication].delegate;
+  if (!appDelegate) {
+    return nil;
+  }
+  if ([appDelegate conformsToProtocol:@protocol(GULMulticastAppDelegateProtocol)]) {
+    id<GULMulticastAppDelegateProtocol> multicastAppDelegate =
+        (id<GULMulticastAppDelegateProtocol>)appDelegate;
+    return multicastAppDelegate;
+  }
+  if (appDelegate && [appDelegate respondsToSelector:@selector(getMulticastDelegate)]) {
+    id<GULMulticastAppDelegateProtocol> multicastDelegate =
+        [appDelegate performSelector:@selector(getMulticastDelegate)];
+    CFRetain((__bridge CFTypeRef)(multicastDelegate));
+    return multicastDelegate;
+  }
+  return nil;
+}
+
+- (id<GULMulticastAppDelegateProtocol>)getMulticastDelegate {
+  return self;
+}
+
+- (void)addInterceptorWithDelegate:(id<GULApplicationDelegate>)interceptor {
+  [_interceptors addObject:interceptor];
+}
+
+- (void)removeInterceptorWithDelegate:(id<GULApplicationDelegate>)interceptor {
+  [_interceptors removeObject:interceptor];
+}
+
+- (BOOL)respondsToSelector:(SEL)aSelector {
+  if ([[self class] instancesRespondToSelector:aSelector]) {
+    return YES;
+  }
+  for (id<GULApplicationDelegate> interceptor in _interceptors) {
+    if (interceptor && [interceptor respondsToSelector:aSelector]) {
+      return YES;
+    }
+  }
```

Comment:
Is there a case where `_interceptors` contains a `id<GULApplicationDelegate>` that does **not** respond to the given `aSelector`?

Reply:
yeah because there are multiple interceptors from auth, messaging and test app and each has different list of selectors.

Reply:
I see now, thanks